scout/server/blueprints/cases/controllers.py

```diff
@@ -817,6 +820,131 @@ def multiqc(store, institute_id, case_name):
+        store.case_collection.update({"_id": case_obj["_id"]}, {"$unset": {"beacon": 1}})
+    else:
+        flash_color = "warning"
+    flash(f"Beacon responded:{resp['message']}", flash_color)
+
+
+def beacon_add(form):
+    """Save variants from one or more case samples to the Beacon server.
+       Handle a POST request to the /apiv1.0/add Beacon endpoint
+
+    Args:
+        form(werkzeug.datastructures.ImmutableMultiDict): beacon submission form
+
+    """
+    if prepare_beacon_req_params() is None:
+        flash(
+            "Please check config file. It should contain both BEACON_URL and BEACON_TOKEN",
+            "warning",
+        )
+        return
+    request_url, req_headers = prepare_beacon_req_params()
+
+    case_obj = store.case(case_id=form.get("case"))
+    # define case individuals (individual_id, same as in VCF) to filter VCF files with
+    individuals = []
+    if form.get("samples") == "affected":
+        individuals = [
+            ind["individual_id"] for ind in case_obj["individuals"] if ind["phenotype"] == 2
+        ]
+    else:
+        individuals = [ind["individual_id"] for ind in case_obj["individuals"]]
+
+    # define genes to filter VCF files with
+    gene_filter = set()
+    for panel in form.getlist("panels"):
+        gene_filter.update(store.panel_to_genes(panel_id=panel, gene_format="hgnc_id"))
+    gene_filter = list(gene_filter)
+
+    submission = {
+        "created_at": datetime.datetime.now(),
+        "user": current_user.email,
+        "samples": individuals,
+        "panels": form.getlist("panels"),
+        "vcf_files": [],
+    }
+
+    # Prepare beacon request data
+    assembly = "GRCh37" if "37" in str(case_obj["genome_build"]) else "GRCh38"
+    data = {
+        "dataset_id": "_".join([case_obj["owner"], assembly]),
+        "samples": individuals,
+        "assemblyId": assembly,
+    }
+    if gene_filter:  # Gene filter is not mandatory
+        data["genes"] = {"ids": gene_filter, "id_type": "HGNC"}
+
+    # loop over selected VCF files and send an add request to Beacon for each one of them
+    vcf_files = form.getlist("vcf_files")
+    if not vcf_files:
+        flash("Please select at least one VCF file to save to Beacon", "warning")
+        return
+    for vcf_key in form.getlist("vcf_files"):
+        data["vcf_path"] = case_obj["vcf_files"].get(vcf_key)
+        resp = post_request_json("/".join([request_url, "add"]), data, req_headers)
+        if resp.get("status_code") != 200:
+            flash(f"Beacon responded:{resp['message']}", "warning")
+            continue
+        submission["vcf_files"].append(vcf_key)
+
+    if len(submission["vcf_files"]) > 0:
+        flash(
+            f"Variants from the following files are going to be saved to Beacon:{submission['vcf_files']}",
+            "success",
+        )
+        store.case_collection.find_one_and_update(
+            {"_id": case_obj["_id"]}, {"$set": {"beacon": submission}}
+        )
+    return
+
+
```

Comment:
This source file is getting really large. Should be refactored some day.

Reply:
Yes!